File: adapters/foxglove/gantry_foxglove/protocol.py
"""Codec for the open ``foxglove.websocket.v1`` protocol (client side).

Spec: https://github.com/foxglove/ws-protocol (docs/spec.md). This module is a
pure, hardware-free codec — no I/O — so it is exercised directly from raw byte
fixtures in the tests. Only the subset a read-only *subscriber* needs is
implemented: parse the server's JSON control messages (serverInfo / advertise /
unadvertise / status), build the client's subscribe / unsubscribe JSON, and
decode the binary server frames (MessageData opcode 0x01, Time opcode 0x02).

Binary framing is little-endian, matching the spec:

    MessageData := 0x01 | subscription_id (u32 LE) | log_time (u64 LE, ns) | payload
    Time        := 0x02 | timestamp (u64 LE, ns)

Message payloads are decoded per the channel's ``encoding``: lerobot scalar
channels use ``json``; its image channels use ``protobuf`` (a
``foxglove.CompressedImage`` message, from which :func:`extract_compressed_image`
pulls the JPEG bytes without a protobuf dependency).
"""

from __future__ import annotations

import json
import struct
from dataclasses import dataclass

# WebSocket subprotocol offered on connect. A server that does not select it is
# not speaking this protocol.
SUBPROTOCOL = "foxglove.websocket.v1"

# Binary server->client opcodes (first byte of a binary frame).
OP_MESSAGE_DATA = 0x01
OP_TIME = 0x02

# JSON server->client op names we act on (others are logged/ignored).
SERVER_INFO = "serverInfo"
ADVERTISE = "advertise"
UNADVERTISE = "unadvertise"
STATUS = "status"


@dataclass(frozen=True)
class Channel:
    """A channel as announced in a server ``advertise`` message."""

    id: int
    topic: str
    encoding: str
    schema_name: str = ""
    schema: str = ""
    schema_encoding: str = ""

    @classmethod
    def from_json(cls, obj: dict) -> "Channel":
        return cls(
            id=int(obj["id"]),
            topic=str(obj["topic"]),
            encoding=str(obj.get("encoding", "")),
            schema_name=str(obj.get("schemaName", "")),
            schema=str(obj.get("schema", "")),
            schema_encoding=str(obj.get("schemaEncoding", "")),
        )


@dataclass(frozen=True)
class MessageData:
    """A decoded binary MessageData frame (opcode 0x01)."""

    subscription_id: int
    log_time: int  # nanoseconds, from the producer's clock
    payload: bytes


@dataclass(frozen=True)
class TimeMessage:
    """A decoded binary Time frame (opcode 0x02)."""

    timestamp: int  # nanoseconds


# --- client -> server JSON messages ---------------------------------------


def encode_subscribe(subscriptions) -> str:
    """Build a ``subscribe`` message.

    ``subscriptions`` is an iterable of ``(subscription_id, channel_id)`` pairs.
    """
    return json.dumps(
        {
            "op": "subscribe",
            "subscriptions": [
                {"id": int(sub_id), "channelId": int(chan_id)} for sub_id, chan_id in subscriptions
            ],
        }
    )


def encode_unsubscribe(subscription_ids) -> str:
    """Build an ``unsubscribe`` message for the given subscription ids."""
    return json.dumps({"op": "unsubscribe", "subscriptionIds": [int(s) for s in subscription_ids]})


# --- server -> client JSON messages ---------------------------------------


def decode_json(text: str) -> dict:
    """Parse a server JSON control message into a dict (must have an ``op``)."""
    obj = json.loads(text)
    if "op" not in obj:
        raise ValueError("foxglove JSON message missing 'op'")
    return obj


def parse_advertise(obj: dict):
    """Return the list of :class:`Channel` in an ``advertise`` message."""
    return [Channel.from_json(c) for c in obj.get("channels", [])]


def parse_unadvertise(obj: dict):
    """Return the list of channel ids in an ``unadvertise`` message."""
    return [int(c) for c in obj.get("channelIds", [])]


# --- server -> client binary messages -------------------------------------


def decode_binary(data: bytes):
    """Decode a binary server frame into a :class:`MessageData` or
    :class:`TimeMessage`. Returns ``None`` for an unknown/empty opcode.

    Raises ``ValueError`` on a truncated frame of a known opcode so a corrupt
    stream is surfaced rather than silently mis-decoded.
    """
    if not data:
        return None
    op = data[0]
    if op == OP_MESSAGE_DATA:
        if len(data) < 13:
            raise ValueError(f"MessageData frame too short: {len(data)} bytes")
        sub_id, log_time = struct.unpack_from("<IQ", data, 1)
        return MessageData(subscription_id=sub_id, log_time=log_time, payload=data[13:])
    if op == OP_TIME:
        if len(data) < 9:
            raise ValueError(f"Time frame too short: {len(data)} bytes")
        (ts,) = struct.unpack_from("<Q", data, 1)
        return TimeMessage(timestamp=ts)
    return None


def encode_message_data(subscription_id: int, log_time: int, payload: bytes) -> bytes:
    """Encode a MessageData frame. Used by the test fake-server (and symmetry)."""
    return b"\x01" + struct.pack("<IQ", int(subscription_id), int(log_time)) + bytes(payload)


# --- protobuf: foxglove.CompressedImage extraction ------------------------
#
# foxglove-sdk logs images on a protobuf channel. We only need two fields of the
# CompressedImage message, so rather than depend on protobuf we scan the wire
# format directly. Field numbers are from the foxglove schema
# (schemas/proto/foxglove/CompressedImage.proto):
#     timestamp = 1 (message), data = 2 (bytes), format = 3 (string),
#     frame_id  = 4 (string)


def _read_varint(buf: bytes, i: int):
    """Read a base-128 varint at offset ``i``; return (value, next_offset)."""
    shift = 0
    result = 0
    while True:
        if i >= len(buf):
            raise ValueError("truncated varint")
        b = buf[i]
        i += 1
        result |= (b & 0x7F) << shift
        if not (b & 0x80):
            return result, i
        shift += 7


def extract_compressed_image(payload: bytes):
    """Extract ``(data, format)`` from a ``foxglove.CompressedImage`` protobuf
    message. ``data`` is the compressed image bytes (e.g. JPEG), ``format`` the
    format string (e.g. ``"jpeg"``). Unknown fields are skipped. Missing fields
    default to ``b""`` / ``""``.
    """
    data = b""
    fmt = ""
    i = 0
    n = len(payload)
    while i < n:
        tag, i = _read_varint(payload, i)
        field = tag >> 3
        wire = tag & 0x07
        if wire == 0:  # varint
            _, i = _read_varint(payload, i)
        elif wire == 1:  # 64-bit
            i += 8
        elif wire == 2:  # length-delimited
            length, i = _read_varint(payload, i)
            chunk = payload[i : i + length]
            i += length
            if field == 2:
                data = bytes(chunk)
            elif field == 3:
                fmt = chunk.decode("utf-8", "replace")
        elif wire == 5:  # 32-bit
            i += 4
        else:
            raise ValueError(f"unsupported protobuf wire type {wire}")
    return data, fmt

File: core/go/foxglove/protocol.go
// Package foxglove is an in-process client for the open
// `foxglove.websocket.v1` protocol (lerobot --display_mode=foxglove, ROS 2
// foxglove_bridge). It is the Go counterpart of the Python reference tap under
// adapters/foxglove: a read-only subscriber that connects, tracks advertised
// channels, decodes binary MessageData frames, maps scalar payloads to Gantry
// frames via a config-driven mapping engine, and hands them to an ingest sink.
//
// This file is the pure, hardware-free codec — no I/O — so it is exercised
// directly from raw byte fixtures. Only the subset a subscriber needs is
// implemented: parse the server's JSON control messages (serverInfo / advertise
// / unadvertise / status), build the client's subscribe / unsubscribe JSON, and
// decode the binary server frames (MessageData opcode 0x01, Time opcode 0x02).
//
// Binary framing is little-endian, matching the spec
// (https://github.com/foxglove/ws-protocol):
//
//	MessageData := 0x01 | subscription_id (u32 LE) | log_time (u64 LE, ns) | payload
//	Time        := 0x02 | timestamp (u64 LE, ns)
//
// Channel encoding: json payloads only in v1 (the scalar topics). Protobuf image
// channels are recognized and skipped by the client (video stays with the Python
// tap; an in-bench video tee would need an h264 encoder — a future
// ffmpeg-managed-binary pattern).
package foxglove

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
)

// Subprotocol is the WebSocket subprotocol offered on connect. A server that
// does not select it is not speaking this protocol.
const Subprotocol = "foxglove.websocket.v1"

// Binary server->client opcodes (first byte of a binary frame).
const (
	opMessageData = 0x01
	opTime        = 0x02
)

// JSON server->client op names the client acts on (others are logged/ignored).
const (
	opAdvertise   = "advertise"
	opUnadvertise = "unadvertise"
	opServerInfo  = "serverInfo"
	opStatus      = "status"
)

// EncodingJSON is the only channel encoding decoded in v1 (scalar topics). Any
// other encoding (e.g. "protobuf" image channels) is recognized and skipped.
const EncodingJSON = "json"

// Channel is a channel as announced in a server `advertise` message. Only the
// fields the client needs are decoded.
type Channel struct {
	ID         int    `json:"id"`
	Topic      string `json:"topic"`
	Encoding   string `json:"encoding"`
	SchemaName string `json:"schemaName"`
}

// MessageData is a decoded binary MessageData frame (opcode 0x01).
type MessageData struct {
	SubscriptionID uint32
	LogTime        uint64 // nanoseconds, from the producer's clock
	Payload        []byte
}

// serverMessage is the union of the server->client JSON control messages we
// parse. Fields not relevant to a given op stay zero.
type serverMessage struct {
	Op         string    `json:"op"`
	Channels   []Channel `json:"channels"`   // advertise
	ChannelIDs []int     `json:"channelIds"` // unadvertise
	Name       string    `json:"name"`       // serverInfo
	Level      int       `json:"level"`      // status
	Message    string    `json:"message"`    // status
}

// decodeJSON parses a server JSON control message. It errors if the message is
// not an object with an `op`, so a malformed control message is surfaced rather
// than silently treated as a no-op.
func decodeJSON(text []byte) (serverMessage, error) {
	var m serverMessage
	if err := json.Unmarshal(text, &m); err != nil {
		return m, fmt.Errorf("foxglove: bad JSON control message: %w", err)
	}
	if m.Op == "" {
		return m, fmt.Errorf("foxglove: JSON control message missing 'op'")
	}
	return m, nil
}

// subEntry is one (subscription_id, channel_id) pair in a subscribe message.
type subEntry struct {
	ID        int `json:"id"`
	ChannelID int `json:"channelId"`
}

// encodeSubscribe builds a `subscribe` message for the given pairs.
func encodeSubscribe(subs []subEntry) ([]byte, error) {
	return json.Marshal(struct {
		Op            string     `json:"op"`
		Subscriptions []subEntry `json:"subscriptions"`
	}{Op: "subscribe", Subscriptions: subs})
}

// encodeUnsubscribe builds an `unsubscribe` message for the given subscription
// ids. Retained for symmetry with the protocol; the client drops subscriptions
// implicitly on unadvertise (a channel that is gone cannot be unsubscribed).
func encodeUnsubscribe(subscriptionIDs []int) ([]byte, error) {
	return json.Marshal(struct {
		Op              string `json:"op"`
		SubscriptionIDs []int  `json:"subscriptionIds"`
	}{Op: "unsubscribe", SubscriptionIDs: subscriptionIDs})
}

// decodeBinary decodes a binary server frame into a *MessageData. Time frames
// (opcode 0x02) and unknown/empty opcodes return (nil, nil) — the subscriber
// does not need them. A truncated frame of a known opcode is an error so a
// corrupt stream is surfaced rather than mis-decoded.
func decodeBinary(data []byte) (*MessageData, error) {
	if len(data) == 0 {
		return nil, nil
	}
	switch data[0] {
	case opMessageData:
		if len(data) < 13 {
			return nil, fmt.Errorf("foxglove: MessageData frame too short: %d bytes", len(data))
		}
		subID := binary.LittleEndian.Uint32(data[1:5])
		logTime := binary.LittleEndian.Uint64(data[5:13])
		// Copy the payload so it does not alias the connection's read buffer,
		// which coder/websocket may reuse on the next Read.
		payload := make([]byte, len(data)-13)
		copy(payload, data[13:])
		return &MessageData{SubscriptionID: subID, LogTime: logTime, Payload: payload}, nil
	case opTime:
		if len(data) < 9 {
			return nil, fmt.Errorf("foxglove: Time frame too short: %d bytes", len(data))
		}
		return nil, nil
	default:
		return nil, nil
	}
}

// encodeMessageData encodes a MessageData frame. Used by the test fake server
// (and for symmetry with the codec).
func encodeMessageData(subscriptionID uint32, logTime uint64, payload []byte) []byte {
	out := make([]byte, 13+len(payload))
	out[0] = opMessageData
	binary.LittleEndian.PutUint32(out[1:5], subscriptionID)
	binary.LittleEndian.PutUint64(out[5:13], logTime)
	copy(out[13:], payload)
	return out
}
